### DISCUSSION TITLE
Klipper @ Kobra Go / Neo

### DISCUSSION THREAD

**Discussion by 1coderookie** [1coderookie/KobraGoNeoInsights#6] [General]
Here we can chat about anything related to [Klipper and the Go/Neo](https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/), just in case anybody wants to ask or mention specific things.  

*Please make sure to read the belonging chapter [Klipper (Mod)](https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/) of my page as well as the [official Klipper documentation](https://www.klipper3d.org/) before asking anything tho.*

**Please keep in mind that I won't answer specific questions about Klipper itself as there's a comprehensive official documentation: https://www.klipper3d.org/**

Compiled `klipper.bin` and preconfigured `printer.cfg` available here: https://github.com/1coderookie/Klipper4KobraGoNeo
**USE AT YOUR OWN RISK!**

**Comment by 1coderookie** [comment]
@SteveBz - So your question was "Interesting about Klipper. I read the link you sent and it says you lose your screen. Is that not a problem?". 
Actually I thought the same initially and that was the only reason which was holding me back switching to Klipper right away at the beginning. Now that I did switch, I can say that I don't really miss the control unit tbh. 

Just as an example: 
- Klipper *keeps* the z-offset, so you don't have to fiddle arount at the control unit for that before each and every print because the stcok-fw forgot your settings once again. 
- Setting temps for e.g. heating up the hotend to change a nozzle or load/unload filament can be done by using Mainsail (also OctoPrint or Fluidd of course, but I prefer using Mainsail, so I'll only refer to that in the following; see the pages why). 
- Initiating the ABL procedure is also be done thru Mainsail, just like everything else.

However, if you really do miss the option to control the printer right at that place later, you have several options to do so:
- You can use your smartphone or tablet to just open the mainsail.local page at your browser.
- You can use an old smartphone or tablet and install KlipperScreen onto it, which then gives you a nice adapted UI with control buttons. I also set it up, printed a holder for it and mounted it where the original control unit was, but actually I unplugged the phone again as I just control it via my computer and the mainsail.local page anyway. It also kinda bothers me that I have to boot up the phone all the time as I don't leave the printer on 24/7. I'll add pictures of it to the pages soon tho.
- If you have a Raspberry Pi running for hosting Mainsail, you can add a touchscreen to that and use KlipperScreen. It's the same UI like using a smartphone or tablet, it's just that you can set up the RPi+screen as 'one' unit in a dedicated case. I'll probably do that, just for the pure comfort of not having to take care about switching the smartphone on and off as the display is powered by the RPi. I just didn't want to spend money again when trying Klipper initially, that's why I used an old smartphone in first place.

**Reply by SteveBz** [reply]
So have you taken the screen off the Neo completely, so that you can re-utilise the space?

**Reply by SteveBz** [reply]
If you wanted to reflash the original, where would you find it?  I guess you need to print out your calibrations, so that youc an re-enter them in Klipper, or when reverting.

**Reply by 1coderookie** [reply]
> So have you taken the screen off the Neo completely, so that you can re-utilise the space?

Exactly, I removed it completely (also the cable) and used that holder to mount the smartphone. I'll add a picture of it tomorrow in the Klipper chapter.

> If you wanted to reflash the original, where would you find it? I guess you need to print out your calibrations, so that youc an re-enter them in Klipper, or when reverting.

Stock firmware: https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_marlin/
Don't know what you mean with the second sentence though. If you wanna go back to stock fw from Klipper, you just flash the stock firmware.bin from Anycubic for the Neo and that's it.

Please read the chapters and the reddit thread I linked to, it's everything described there imho.
EDIT: I just updated the description https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/#installation so maybe it becomes more clear now..

**Reply by SteveBz** [reply]
> > I guess you need to print out your calibrations, so that you an re-enter them in Klipper, or when reverting.
> 
I meant the stock calibrations are (from memory) X80 Y80 Z80 E400.  My calibrated values are slightly different. That's all.

**Reply by 1coderookie** [reply]
Ah ok, got it. Well, sure you can do that, but you'd have to do the calibrations anyway again and you have to calculate and set the value for the rotation distance instead of the steps in Klipper: https://www.klipper3d.org/Rotation_Distance.html
The values are preconfigured in the mentioned cfg-files on the reddit post, so you can start with that as well..

**Comment by SteveBz** [comment]
And you don't need to burn a bootloader, you just boot Klipper from the SD card. 

Would a bootloader be better? Did you try? 

Steve. 

⁣Get BlueMail for Android ​

**Reply by 1coderookie** [reply]
What do you mean by bootloader? To the mainboard? No. 
You have to create and **flash** the `firmware.bin` file to the MCU/mainboard as described in the reddit post I linked to (and like you'd do if you'd update the stock firmware). 
Then you'll need a host system (like a RPi, ThinClient, old laptop or so) to install e.g. Mainsail for accessing and configuring Klipper. This is necessary as it communicates with the Klipper firmware of your mainboard, you won't be able to use the printer without it.

**Reply by SteveBz** [reply]
This is what I mean by flashing a bootloader.  It means you can reflash direct from your Arduino IDE over a USB cable.

https://letsprint3d.net/guide-how-to-flash-a-bootloader-on-melzi-boards/

**Reply by 1coderookie** [reply]
No need to mess around with the mainboard. Iirc the MCU isn't supported by ArduinoIDE anyway and you're creating the necessary firmware.bin-file of Klipper as how it's described in the official documentation I linked to from the page.  

As I already said and how it's described at my pages and in the thread: compile the bin-file following the instructions (note the mentioned settings for MCU and serial in the Installation chapter and/or the reddit post), flash Klipper's firmware.bin from mSD card, take it out and then connect the printer with Klipper to your RPi which has OctoPrint/Mainsail/.. running. 

Same procedure as when you wanna go back to stock fw: just flash it using the mSD card and you're good as described here https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_marlin/#update-procedure  

Again: please read the chapters and follow the links, I can't and don't want to describe everything again here as it's already explained at the pages..

**Comment by SteveBz** [comment]
As you know I currently use OctoPi on Rpi.  Presumably I can just continue with this, or should I switch to MainSail.  Why did you switch?
Steve.

**Reply by 1coderookie** [reply]
Yes, you could continue using OctoPrint, that's fine and no problem, it's even the software they refer to on their documentation page when it comes down to checking the basic functions and showing screenshots of it.  
However, I personally switched to Mainsail with using Klipper as it's pretty much tailored for Klipper. With Octo for example you'd need to SSH into the Pi for uploading your files like the printer.cfg, also if you need to change settings (which you'll have to) and you want to edit that file, you'd have to do so on your computer and then SHH into the Pi again and upload it. Maybe there's also a plugin for that in Octo, but for me that was too much of a hassle. Mainsail does have an inbuilt editor which makes everything just super easy and comfortable. You can also create and edit macros there, already have the heightmap function inbuilt, the tabs of Octo (which always bothered me) are gone and so on. You can also use thumbnail previes of your model files if you activate that in the slicer (dunno if Cura can do that tho, but SuperSlicer does). 
As I said, it's tailored for Klipper usage, so imho the best solution. (Not to mention the much nicer look&feel of it imho.)
I wrote a bit about it here: https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/#octoprint-mainsail-or-fluidd  
I'll probably add screenshots and some specific tips for Mainsail&SuperSlicer there as well soon..
If you have another mSD card, just install MainsailOS using RaspberryPiImager and give it a try - I'd bet you'll like and prefer it.. ;) If you don't like it (what I'd doubt), swap out the cards and you're back to Octo..

**Reply by SteveBz** [reply]
> If you have another mSD card, just install MainsailOS using RaspberryPiImager and give it a try - I'd bet you'll like and prefer it.. ;) If you don't like it (what I'd doubt), swap out the cards and you're back to Octo..

Great idea.  I have several.

**Reply by SteveBz** [reply]
So I created a Mainsail build just from Raspberry Pi Imager and it worked perfectly.  Just used the imager to set up WiFi and everything.  So that was amazingly easy and just worked.  Now Klipper.....

**Comment by SteveBz** [comment]
I can't find a simple description of the Klipper firmware install.  xpeng121 has obviously managed it, but I can't see what he's doing:

1) git clone https://github.com/Klipper3d/klipper
(I don't have git on my windows machine, but I do on a linux machine so I flicked through to Linux to do this.  Maybe step 0, install git on Windows like this <insert link>)
2) Compile firmware: Pls follow the klipper's official tutorial and used the following config.  
(No link. Is it 'make' or something?)
Change mcu to hc32f460 and the serial to Anycubic Kobra
(where is this?  Before or after the compile?)

And so on ....

These instructions need background knowledge.

Sorry to be so negative.

**Reply by 1coderookie** [reply]
If you have to criticize the post, please do so at reddit, not here. I'm not responsible for the post. 

Besides that, it really seems like you don't even try to read carefully (neither the whole thread of that post from xpeng121 nor my pages) or even try to use Gooogle..
 
I linked to the installation docs of Klipper https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/#installation in the ***fourth*** sentence: 
 
![Bildschirmfoto von 2023-03-08 09-18-59](https://user-images.githubusercontent.com/25869362/223659600-cdbdb9d5-7a1a-4593-8334-8a164cb99ca4.png)


Also Google puts it out in big letter as ***first*** search result by just looking for "klipper":
![Bildschirmfoto von 2023-03-08 09-17-23](https://user-images.githubusercontent.com/25869362/223659290-8bce17c0-a7a7-47fa-9ab6-73aa7f06e573.png)

 
Also in the thread itself the links are mentioned in the ***second*** and ***third*** answer right below the initial post:
![Bildschirmfoto von 2023-03-08 09-09-56](https://user-images.githubusercontent.com/25869362/223657257-2d33da67-7d94-45ce-bcf4-4e8e129b8298.png)


There are also tons of YouTube videos and articles as well about how to compile Klipper firmware in general.

However, even tho my pages aren't about repeating the official installation guide or about how to set up/adjust the `printer.cfg` later  (so don't expect that, neither at my pages nor here btw..), I'll probably add a more detailed description when I find the time to do so. 

So if you can't or don't want to invest some time getting yourself informed about the general compilation process, you probably have to ask at the reddit post or wait until/if I'll add a more detailed step-by-step description at my page and continue using the stock firmware. But also then you'd still have to read attentively - I'm afraid you won't get around that. No offense.   

And just as a kind 'warning' (even tho it's already mentioned at my page as well): 
Klipper isn't like "copy that cfg-file and start printing" - you'd *have* to check the settings and adjust them to your specific machine and needs. 
E.g. when you want to adjust the e-steps, you'd also have to do this using a formula to get the *rotation distance* which involves reading the official documentation as well. 
So if that's nothing for you and you don't have any problems right now using the stock firmware, you probably shouldn't switch to Klipper then..

**Reply by SteveBz** [reply]
Hi Ulf, thanks for responding.  Sorry.

You will have a detailed description in a day or two, as soon as I get through this. 

>  tons of YouTube videos and articles

... are not really my thing.  

A nice clear list is.

**Reply by 1coderookie** [reply]
It's ok, it's just that I can't take the "effort of reading" off of you so to say and the procedure is described at the official documentation which is linked to and about which you complained to not have any links.   

Thanks for the offer about the description, but I'll do it myself tho (also here: don't get me wrong, no offense, I appreciate the offer - it's just that I prefer doing something like this on my own ;) ).

I just created a repo where you can find the compiled file soon, in case you prefer to not compile it by your own. It's the file I flashed my Neo with and it works, but keep in mind that ***you'd use it at your own risk***. So if you prefer to compile it on your own to be sure that you did everything right and don't wanna risk using an already compiled file, it's fine of yourse.. 
Here's the link: https://github.com/1coderookie/Klipper4KobraGoNeo - maybe check it out later as I don't have time right now to write and upload the stuff, hopefully will get it done soon tho..

**Reply by 1coderookie** [reply]
Update: 
Files `klipper.bin` added, `printer.cfg` also added. 
Other `.cfg` files like `gcode.cfg` not yet but you can find them at the mentioned thread. 

As I said before: use at your own risk tho..

**Comment by SteveBz** [comment]
Hi Ulf,

So the main issue for me was that I didn't understand that I was working out of the RPi.  I just thought that:
1) Prepare a mainsail SD from Raspberry Pi Imager;
2) Compile and prepare a boot disk from from your local PC
3) Boot the Neo.

But I didn't realise that 2) meant prepare a boot disk from your Mainsail or OctoPi machine NOT your Windows PC (I don't know how this story ends yet so that was the source of my frustration this morning).  Anyhow not to worry.  That point could be clearer.  So that's where I'm working now.

Secondly the various config files you and others off all differ a little:

- They have different include file;
- They have different max and min values.  

Does this matter?

**Reply by 1coderookie** [reply]
You need to build it from that computer where you downloaded the Klipper repo for being able to run the "make menuconfig" out of the klipper directory - imho that's pretty clear in the official documentation.

About the slightly different printer.cfg files: as mentioned you need to work yourself trough the configuration of the file as it needs to fit your machine and needs. Even though they're the same, some wants to use different spots for the ABL, different z-offsets, different whatever. In my file you'll also finde the "screws_tilt_calibrate" section (at pretty much the end of the file) for manual bed leveling which I do in addition to the ABL - others don't..  
Same goes for the include files - I personally added some I'm using , others didn't. Again you also need to check and decide what you want, have or generate by yourself and therefore want to include that. It's all explained at the official doc..

**Comment by EmeldaMentnechJ** [comment]
It's great to have a dedicated resource for Klipper and the Go/Neo! I appreciate the links to the official Klipper documentation and your page's Klipper (Mod) chapter for additional information. Thanks for also providing the compiled klipper.bin and preconfigured printer.cfg - it's always good to have a starting point when working with new software. However, it's important to note that using third-party software carries some risk, so it's important to exercise caution and thoroughly research any changes or modifications before implementing them.

**Comment by ronaldgrant197** [comment]
It seems like you're inviting people to discuss Klipper and Go/Neo-related topics, while also directing them to read the relevant chapter on your page and the official Klipper documentation. You also offer compiled klipper.bin and preconfigured printer.cfg files on your GitHub page, but with a disclaimer to use them at one's own risk.

**Comment by ShontaLazariniw** [comment]
It's great to see that you're offering resources for people interested in Klipper and the Go/Neo. It's important to emphasize the need for users to read the relevant documentation before asking questions, as that can help prevent unnecessary inquiries and ensure that the answers provided are accurate. Additionally, it's important to make it clear that any compiled binaries or preconfigured configurations should be used at the user's own risk, as they may not be suitable for everyone's individual needs or configurations.

**Comment by LorinaMalmquistAwj** [comment]
Hello! It sounds like you're interested in chatting about Klipper and the Go/Neo, and you've provided some helpful resources for others to reference.

For those who may not be familiar, Klipper is a 3D printer firmware that runs on a single-board computer like the Raspberry Pi, and it provides advanced features and performance improvements over traditional firmware. The Go/Neo is a 3D printer that is compatible with Klipper and provides high-quality prints at an affordable price point.

It's great that you've provided a warning about using the compiled klipper.bin and preconfigured printer.cfg from your GitHub repository at one's own risk. It's important to always be cautious when using third-party software and to thoroughly research and understand any potential risks before using it.

For anyone who is interested in learning more about Klipper and its capabilities, it's recommended to read the official Klipper documentation and the belonging chapter on your page. Thank you for sharing your knowledge and resources with the community!

**Comment by TerrenceTiank** [comment]
As an AI language model, I don't have personal experience or knowledge about Klipper, Go/Neo or 3D printing, but I'm happy to help answer any specific questions you may have regarding these topics based on my training data and available knowledge. Feel free to ask

**Comment by SteveBz** [comment]
Hi Ulf,

I promised to get back to you when I had Klipper working.  I have the install finshished, but not the printing yet.  This what I consider the install to be:

- Create Marlin SD for reverting to original (https://cdn.shopify.com/s/files/1/0245/5519/2380/files/Kobra_Neo_v1.3.3_firmware.rar?v=1675912197). And unzip the .rar file.
- Install standard Mainsail on RPi3b+ with Raspberry Installer.
- Insert SD on RPi and reboot pi.
- Navigate to Mainsail Dashboard in Browser to access RPi (eg (192.168.1 nnn)
- Copy and modify [u/codeneobee](https://www.reddit.com/u/codeneobee/)'s printer.cfg. (Comment out calibration.cfg)
- Copy [u/xpeng121](https://www.reddit.com/u/xpeng121/)'s gcode.cfg from pastebin too.
- Now you get an error in Mainsail that "Option 'control' in section 'heater_bed' must be specified". This will fix that (in the right sections respectively). Copied from codeneobee again here:
`        
       [extruder]
        control = pid
        pid_kp = 18.247
        pid_ki = 0.929
        pid_kd = 89.640

        [heater_bed]
        control = pid
        pid_kp = 59.562
        pid_ki = 1.013
        pid_kd = 875.5

        [stepper_z]
        position_endstop: 0
        rotation_distance: 8
`
- Download the klipper firmware.bin from [u/codeneobee](https://www.reddit.com/u/codeneobee/) to SD card, insert in printer and reboot.
- Verify temperatures - all works (Now second time through - still works).
- Sort out homing with all the X, Y & Z figures needing adjusting. Homing works
- Now verify temperatures doesn't work on either bed or extruder. I can't imagine what happened. (reverted - everything's OK. Start again.)
- OK. Seems to be working now. Completed https://www.klipper3d.org/Config_checks.html

However, there are a huge number of configuration checks that I haven't yet completed.

Kind regards

**Reply by 1coderookie** [reply]
Congrats that you made it :)

> * Now you get an error in Mainsail that "Option 'control' in section 'heater_bed' must be specified". This will fix that (in the right sections respectively). Copied from codeneobee again here:
>   `
>   [extruder]
>   control = pid
>   pid_kp = 18.247
>   pid_ki = 0.929
>   pid_kd = 89.640
>   ```
>     [heater_bed]
>     control = pid
>     pid_kp = 59.562
>     pid_ki = 1.013
>     pid_kd = 875.5
>   

Dunno why you got that error message actually as I don't have it configured in the 'manual' section and didn't get it, in my config it's in the auto-generated section at the very end. That's at least the place to be imho for the values, as they'll be changed and updated whenever you do a PID tune for the bed and save it to the config.

**Comment by ptrstourac** [comment]
Hi!
I finally Klipperized my Kobra Go after just another rant about stock-fw forgetting my Z-offset.

Dunno if it's just me, but time to time my E-stepper just shuts off midprint (doesn't grind filament etc., disables itself). 
I've read there may be an issue with TMC2208 drivers and pressure advance on Klipper GH, but since I saw it configured in some configs for Kobra Go, I wonder if anyone has encountered this behavior too.

printer.cfg: https://pastebin.com/f4cwJ77B

Thanks for any suggestions

**Reply by 1coderookie** [reply]
Hi,
hmm, that's weird and I'm afraid I can't help you with that. This 'stalling' is a known issue when trying to use the stock firmware (Marlin) with advanced Linear Advance and that's actually one of the most common reasons why people switch to Klipper when using the Go/Neo. But until now I didn't hear anyone complaining about this problem when using the Go/Neo with PA. 
I'm using the Neo (which uses the same mainboard as the Go) with Pressure Advance and never had any problems.  
Maybe @xiaopeng12138 knows anything about it, as he is using the Go and also created the printer.cfg for the Klipper repository?

**Reply by xiaopeng12138** [reply]
what is your e-stepper temperature? if it's way too hot, the driver will also stop working. it should be stay at a temp that can be touch with hand for a while.

**Reply by 1coderookie** [reply]
That's a good idea! Is your mainboard fan working @ptrstourac ?

**Reply by ptrstourac** [reply]
Sorry for taking so long to reply (exam period on my university is kinda exhausting). From my tests, it seems to happen when running printer on higher speeds (like 70-80 mm/s). Steppers are running cool and mainboard fan works too. 

I wasn´t able to catch the exact moment when this issue happens, but I suppose it happens when extruder looses a step, so it just disables. Tried loosening the idler, even to that level I had problem with under-extrusion. Still had this problem.

Wonder if tinkering with E-stepper current might solve this issue. Since my Y-axis was losing steps even when having the bed belt and wheels tightened correctly, I had to increase current for this particular stepper.

**Reply by 1coderookie** [reply]
I'd be surprised if it's really the vref actually, but I also never heard that anyone had to increase vref for the y-axis yet.  
I assume you triple-checked the connections and the wiring, right? 
Which material and at which temp are you printing? Afaik, printing too cold can also cause skipping steps.  
Did you maybe also check the feeder gear, if everything is correctly assembled and still ok there?

**Reply by xiaopeng12138** [reply]
> Sorry for taking so long to reply (exam period on my university is kinda exhausting). From my tests, it seems to happen when running printer on higher speeds (like 70-80 mm/s). Steppers are running cool and mainboard fan works too. 
> 
> I wasn´t able to catch the exact moment when this issue happens, but I suppose it happens when extruder looses a step, so it just disables. Tried loosening the idler, even to that level I had problem with under-extrusion. Still had this problem.
> 
> Wonder if tinkering with E-stepper current might solve this issue. Since my Y-axis was losing steps even when having the bed belt and wheels tightened correctly, I had to increase current for this particular stepper. 

You can try to check the vref by using the multimeter to measure the voltage between potentiometer housing and ground if you are comfortable with that. 
For me my stock vref for all drivers is around 0.72v. This should be able to print at 2k accel and about 100mm/s. 
I bump my vref now to 0.9v for the moment axis and 1.1v for the extruder because this is the highest my stepper can handle with a reasonable temp. 
I think 0.8v will be a great start point if you want to mess around with. But just be careful! A bad vref can kill your driver and stepper! You should read some article and watch some videos about how to adjust vref on TMC2208.

**Reply by ptrstourac** [reply]
Yes, i had everything checked more than twice. Happens when printing common materials (PLA/PETG), even with PA. Bumping the temperatures was one of my first troubleshooting steps, but still, not a solution to this problem.

I will see what results tinkering with the Vref yields, already changed it to 0.85V from stock 0.7V, but haven't had time to do some test prints. Since I have quite a lot of experience with it, so don't worry i'll destroy something. ;-)

I'll let you know when i print something with newly set Vref.

**Comment by NuclearPhoenixx** [comment]
As far as I understood the differences between Klipper and Marlin, Klipper gets you a better print quality as well, right? How much better will prints be with Klipper?

Do you have a comparison by chance?

**Reply by 1coderookie** [reply]
Well, it's not that Klipper gives you better results automagically without calibration and tuning your settings.   
You can get pretty perfect as well as pretty messed up prints with either fw.  
BUT: with Klipper you can take advantage of using Pressure Advance and Input Shaping, which -if tuned in correctly- drastically improves the quality as well as allows you to print with higher speeds.

**Reply by NuclearPhoenixx** [reply]
Ok, right I see. So sad that Linear Advance doesn't work in the current Marlin FW. I thought that the extra processing power makes the printer somehow more responsive and precise or accurate and therefore you notice a slight increase in quality without much doing. But what you say definitely makes sense too 👍🏻

**Reply by 1coderookie** [reply]
It's really not a big of a hassle tho, it just might look a bit intimidating at first sight. But you can use the files I'm offering in that specific repo so you don't have to build the bin file and just check it out. Really, you should give it a try - once you get used to how it works, how to adjust settings in the printer.cfg and so on it's really handy.  
Use Mainsail or Fluidd instead of OctoPrint for more comfort tho.   

If you really don't like it in the end (which I'd doubt ;) ), then you can still go back to stock fw by just re-flashing it - but then you have to deal with z-offsets that will be forgotton and unreliable again and soon you'll miss Klipper ;D

**Reply by NuclearPhoenixx** [reply]
Would Klipper still be "better" than Marlin if Linear Advance and Input Shaping worked in the Marlin build? I'm going to have a look at the general differences between these two...

**Reply by 1coderookie** [reply]
Well, I never dealt with a Marlin version that is actually capable of it, so I can't say a lot about it.  
But: Klipper still allows you to adjust the settings in the firmware by just adjusting the printer.cfg without the need of recompiling the bin and flashing the mobo every time, so that's a plus already.  
Besides that, Klipper works kinda different than Marlin, as it processes the file on the host and then just sends the commands to the MCU of the printer instead of processing the file on the printer's mobo. But tbh, I don't remember all that exactly anymore and would have to do further research again as well..

**Reply by NuclearPhoenixx** [reply]
Sounds pretty cool, especially being able to modify the config "OTG" instead of having to recompile everything. I'll have a look at it when I've got time again.

**Reply by 1coderookie** [reply]
Yes, it really is. But use Mainsail (or Fluidd), don't go with OctoPrint when using Klipper ;)

Btw: Thanks again for the donation yesterday - I immediately spent it on longer spacers this morning for raising my bed on the K2Pro for adding the insulation, so no treat for the cats this time :smile_cat:

**Reply by NuclearPhoenixx** [reply]
You're welcome 😺

**Comment by JTSJohnny** [comment]
@1coderookie Hello, again!
I am Windows12 from the forum. I am getting Klipper up and running on my thing, using KlipperScreen and mainsail. I still have to do the Manual Z probe. Do I need to set the gcode Z offset? If so, how?

**Reply by NuclearPhoenixx** [reply]
You're on Windows 12?? 😆

**Reply by JTSJohnny** [reply]
> You're on Windows 12?? 😆

No, I use 11, but I am intristed in the windows 12 conwepts, so my name on the klipper forum is windows12

**Reply by NuclearPhoenixx** [reply]
Ooooh my god, I'm sorry. Totally misread that 🤦🏻‍♂️ 😄

**Reply by 1coderookie** [reply]
You need to set the z-offset for both the endstop (which will be z-offset of the actual 1st layer then) and the ABL sensor (which is only to show an accurate mesh etc). Read this chapter https://1coderookie.github.io/KobraGoNeoInsights/firmware/fw_klipper/#z-offset and set both offsets by following the Klipper docs.  
If you then adjust the z-offset for the 1st layer by using additional g-code or by setting it in the firmware is up to you. I personally always do it via MainsailUI: adjust the height using the buttons while printing and when done, save that to the *endstop*.
